official/vision/beta/projects/yolo/modeling/factory.py

```diff
@@ -0,0 +1,148 @@
+from yolo.modeling.heads.yolo_head import YoloHead
+from yolo.modeling.layers.detection_generator import YoloLayer
+from yolo.modeling.backbones.darknet import build_darknet
+
+from yolo.modeling import yolo_model
+from yolo.configs import yolo
+
+
+def build_yolo_decoder(input_specs, model_config: yolo.Yolo, l2_regularization):
+  activation = (
+      model_config.decoder.activation
+      if model_config.decoder.activation != "same" else
+      model_config.norm_activation.activation)
+  subdivisions = 1
+
+  if model_config.decoder.version is None:  # custom yolo
+    model = YoloDecoder(
+        input_specs,
+        embed_spp=model_config.decoder.embed_spp,
+        use_fpn=model_config.decoder.use_fpn,
+        fpn_depth=model_config.decoder.fpn_depth,
+        path_process_len=model_config.decoder.path_process_len,
+        max_level_process_len=model_config.decoder.max_level_process_len,
+        xy_exponential=model_config.decoder.xy_exponential,
+        activation=activation,
+        subdivisions=subdivisions,
+        use_spatial_attention=model_config.use_sam,
+        use_sync_bn=model_config.norm_activation.use_sync_bn,
+        norm_momentum=model_config.norm_activation.norm_momentum,
+        norm_epsilon=model_config.norm_activation.norm_epsilon,
+        kernel_regularizer=l2_regularization)
+    return model
+
+  if model_config.decoder.type == None:
+    model_config.decoder.type = "regular"
+
+  if model_config.decoder.version not in yolo_model.YOLO_MODELS.keys():
+    raise Exception(
+        "unsupported model version please select from {v3, v4}, \n\n \
+        or specify a custom decoder config using YoloDecoder in you yaml")
+
+  if model_config.decoder.type not in yolo_model.YOLO_MODELS[
+      model_config.decoder.version].keys():
+    raise Exception("unsupported model type please select from \
+        {yolo_model.YOLO_MODELS[model_config.decoder.version].keys()},\
+        \n\n or specify a custom decoder config using YoloDecoder in you yaml")
+
+  base_model = yolo_model.YOLO_MODELS[model_config.decoder.version][
+      model_config.decoder.type]
+
+  cfg_dict = model_config.decoder.as_dict()
+  for key in base_model:
+    if cfg_dict[key] is not None:
+      base_model[key] = cfg_dict[key]
+
+  base_dict = dict(
+      activation=activation,
+      subdivisions=subdivisions,
+      use_spatial_attention=model_config.decoder.use_spatial_attention,
+      use_sync_bn=model_config.norm_activation.use_sync_bn,
+      norm_momentum=model_config.norm_activation.norm_momentum,
+      norm_epsilon=model_config.norm_activation.norm_epsilon,
+      kernel_regularizer=l2_regularization)
+
+  base_model.update(base_dict)
+  print(base_model)
+
+  model = YoloDecoder(input_specs, **base_model)
+
+  return model
+
+
+def build_yolo_filter(model_config: yolo.Yolo, decoder: YoloDecoder, masks,
+                      xy_scales, path_scales):
+
+  def _build(values):
```

Comment:
what does this do exactly?

Reply:
added a comment to explain

Reply:
I still feel this logic is a bit obscure, can you please think of a simpler logic?

Reply:
If you can't find a different way to do it. please add an example config to clarify the behavior.

Reply:
Don't use lvls, cfg in description, instead use levels, config, .. etc.